feat(ngMocks): add expectRoute and whenRoute shortcuts with colon param matching

dmaloneycalu opened the pull request:
Adds a `whenRoute()` and `expectRoute()` function to `$httpBackend` service to make mocking out responses similar to writing API routes on a server. Also includes colon delimited param matching so that expect and when handlers can match requests in the form of `/user/:id`, identical to the way ngRoute matches urls in an Angular application.

Here's an [example file](https://gist.github.com/dmaloneycalu/cb258e2f3b987fd687b2) that reflects one of the use cases for the feature (paged API responses).

dmaloneycalu commented:
CLA signed and builds passing. Any feedback here?

dmaloneycalu commented:
Any thoughts or suggestions on this? Potentially a 1.4.x or 1.5 addition?

@vojtajina @petebacondarwin @matsko

dmaloneycalu commented:
Have not seen much action on the Angular repo lately, wanted to give it a bump

dmaloneycalu commented:
I can rebase this on master and squash the commits if necessary. Still haven't received any feedback. @angular

petebacondarwin commented:
I like the look of this @dmaloneycalu please do rebase. We probably won't be able to get to it until after AngularConnect now.

gkalpak commented:
The tests might also make good use of the [`they()`][1] helper.

[1]: https://github.com/angular/angular.js/blob/291d7c467fba51a9cb89cbeee62202d51fe64b09/test/helpers/privateMocks.js#L22

gkalpak commented:
BTW, the real benefit I see in this change comes from the added `keys` parameter. (Matching routes could be achieved with RegExps before.)

So, we need docs and tests for that :)

dmaloneycalu commented:
@petebacondarwin can do.

@gkalpak I'll add additional tests for the parameter addition and rewrite some of the tests to take advantage of the `they()` helper

dmaloneycalu commented:
@petebacondarwin @gkalpak 

- squashed commits
- added additional tests for params matching outside of `whenRoute` and `expectRoute`
- fixed a few documentation lines I was missing

petebacondarwin commented on src/ngMock/angular-mocks.js:
```diff
@@ -1251,20 +1251,21 @@ function createHttpBackendMock($rootScope, $timeout, $delegate, $browser) {
    *   data string and returns true if the data is as expected.
    * @param {(Object|function(Object))=} headers HTTP headers or function that receives http header
    *   object and returns true if the headers match the current definition.
+   * @param {(Array)=} keys Array of keys to assign to regex matches in the respond params for the url.
```
I think that this needs a bit more explanation for people coming to this documentation without looking at the code.  Perhaps we could have a single section in the ngMock.$httpBackend description that explains what can now be done with regexes and keys then simply refer each fo these parameter descriptions to that section?

dmaloneycalu replied:
I'll take a stab at this and notify you when I've pushed it.

dmaloneycalu committed:
feat(ngMocks): add expectRoute and whenRoute shortcuts with colon param matching

Add whenRoute method to $httpBackend to support matching similar to API backends, which
handles responses based on the query to a particular route. Also includes params in the respond
function to give access to both query params and route colon delimited values.

This is a feature request and is not associated with any open issue

dmaloneycalu commented:
@petebacondarwin
- Added a section to the documentation
- Reordered code so documentation for `whenRoute` and `expectRoute` were at bottom of shortcuts list
- Pointed method param declarations to see above for further explanation

I'm not an English professor, so feel free to let me know if anything isn't well defined enough

petebacondarwin commented:
Thanks @dmaloneycalu I will take a look

gkalpak commented on src/ngMock/angular-mocks.js:
````diff
@@ -1125,7 +1125,76 @@ angular.mock.dump = function(object) {
+      });
+  ```
+ *
+ * By default, query parameters on requested urls are parsed onto the params object, so a requsted url
````
_requsted_  -->  _requested_

gkalpak commented on src/ngMock/angular-mocks.js:
```diff
@@ -1125,7 +1125,76 @@ angular.mock.dump = function(object) {
+ * of `/list?q=searchstr&orderby=-name` would set params to be `{q: 'searchstr', orderby: '-name'}`
+ *
+ * For extra convenience, `whenRoute` and `expectRoute` shortcuts are available. These methods offer colon
+ * deliminated matching of solely the requrest url path, ignoring the query string. This allows declarations
```
_deliminated_  -->  _delimited_

gkalpak commented on test/ngMock/angular-mocksSpec.js:
```diff
@@ -963,6 +963,11 @@ describe('ngMock', function() {
 
+    it('should provide "route" shortcuts for expect and when', function() {
+        expect(typeof hb.whenRoute).toBe("function");
+        expect(typeof hb.expectRoute).toBe("function");
```
Too much indentation.

petebacondarwin commented:
I did a bunch of tidy up on the docs and landed. Thanks @dmaloneycalu 
@gkalpak I missed that indentation. Sorry.

dmaloneycalu commented:
Sorry on the indentation, was trying to be thorough checking since my editor is usually 4 spaces for indentation. Glad I could contribute.

gkalpak commented:
:+1:
